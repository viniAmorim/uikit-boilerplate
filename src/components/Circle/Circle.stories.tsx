import React from 'react';

import { Story, Meta } from '@storybook/react';

import { Circle, CircleProps } from './Circle';

export default {
  title: 'Atoms/Circle',
  component: Circle,
} as Meta;

const Template: Story<CircleProps> = (args) => <Circle {...args} />;

export const Default = Template.bind({});
Default.args = {
  type: 'success',
};
