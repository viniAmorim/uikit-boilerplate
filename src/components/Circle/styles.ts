import styled, { css } from 'styled-components';

export type StyledCircleProps = {
  circleColor?: 'green' | 'yellow' | 'red';
};

const colorModifier = {
  green: css`
    background-color: ${({ theme }) => theme.colors.green500};
  `,
  yellow: css`
    background-color: ${({ theme }) => theme.colors.yellow500};
  `,
  red: css`
    background-color: ${({ theme }) => theme.colors.red500};
  `,
};
export const Circle = styled.div<StyledCircleProps>`
  width: 30px;
  height: 30px;
  border-radius: 50%;
  display: flex;
  justify-content: center;
  align-items: center;
  ${({ circleColor = 'green' }) => colorModifier[circleColor]}
`;
