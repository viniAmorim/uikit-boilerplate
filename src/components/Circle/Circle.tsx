/* eslint-disable no-nested-ternary */
import React, { forwardRef } from 'react';

import { Check, Reload, X } from 'tabler-icons-react';

import * as Styled from './styles';

export interface CircleProps
  extends React.HTMLAttributes<HTMLDivElement>,
    Partial<Styled.StyledCircleProps> {
  type: 'success' | 'awaiting' | 'error';
}

const Icon = ({ type }) => {
  switch (type) {
    case 'success':
      return <Check size={20} strokeWidth={2} color="#fff" />;
    case 'awaiting':
      return <Reload size={20} strokeWidth={2} color="#fff" />;
    default:
      return <X size={20} strokeWidth={2} color="#fff" />;
  }
};

export const Circle = forwardRef<HTMLDivElement, CircleProps>(function Circle(
  { type = 'success', ...rest },
  ref,
) {
  const getColor = (color: string): Styled.StyledCircleProps['circleColor'] => {
    switch (color) {
      case 'success':
        return 'green';
      case 'awaiting':
        return 'yellow';
      default:
        return 'red';
    }
  };

  return (
    <Styled.Circle ref={ref} circleColor={getColor(type)} {...rest}>
      <Icon type={type} />
    </Styled.Circle>
  );
});

export default Circle;
