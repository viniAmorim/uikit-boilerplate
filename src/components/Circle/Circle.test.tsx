import React from 'react';

import { render } from '~/test/utils';

import { Circle } from './Circle';

describe('Circle', () => {
  it('should render correctly', () => {
    const { container } = render(<Circle type="success" />);

    const circleElement = container.firstChild as HTMLElement;

    expect(circleElement).toBeInTheDocument();
    expect(circleElement).toMatchSnapshot();
  });

  it('should render an awaiting circle', () => {
    const { container } = render(<Circle type="awaiting" />);

    const circleElement = container.firstChild as HTMLElement;

    expect(circleElement).toBeInTheDocument();
    expect(circleElement).toMatchSnapshot();
  });

  it('should render an error button', () => {
    const { container } = render(<Circle type="error" />);

    const circleElement = container.firstChild as HTMLElement;

    expect(circleElement).toBeInTheDocument();
    expect(circleElement).toMatchSnapshot();
  });
});
