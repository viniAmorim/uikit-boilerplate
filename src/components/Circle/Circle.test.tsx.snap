// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Circle should render an awaiting circle 1`] = `
<div
  class="sc-beqWaB kDavFb"
>
  <svg
    class="icon icon-tabler icon-tabler-reload"
    fill="none"
    height="20"
    stroke="#fff"
    stroke-linecap="round"
    stroke-linejoin="round"
    stroke-width="2"
    viewBox="0 0 24 24"
    width="20"
    xmlns="http://www.w3.org/2000/svg"
  >
    <path
      d="M0 0h24v24H0z"
      fill="none"
      stroke="none"
    />
    <path
      d="M19.933 13.041a8 8 0 1 1 -9.925 -8.788c3.899 -1.002 7.935 1.007 9.425 4.747"
    />
    <path
      d="M20 4v5h-5"
    />
  </svg>
</div>
`;

exports[`Circle should render an error button 1`] = `
<div
  class="sc-beqWaB iDnEgm"
>
  <svg
    class="icon icon-tabler icon-tabler-x"
    fill="none"
    height="20"
    stroke="#fff"
    stroke-linecap="round"
    stroke-linejoin="round"
    stroke-width="2"
    viewBox="0 0 24 24"
    width="20"
    xmlns="http://www.w3.org/2000/svg"
  >
    <path
      d="M0 0h24v24H0z"
      fill="none"
      stroke="none"
    />
    <line
      x1="18"
      x2="6"
      y1="6"
      y2="18"
    />
    <line
      x1="6"
      x2="18"
      y1="6"
      y2="18"
    />
  </svg>
</div>
`;

exports[`Circle should render correctly 1`] = `
<div
  class="sc-beqWaB eKXEiF"
>
  <svg
    class="icon icon-tabler icon-tabler-check"
    fill="none"
    height="20"
    stroke="#fff"
    stroke-linecap="round"
    stroke-linejoin="round"
    stroke-width="2"
    viewBox="0 0 24 24"
    width="20"
    xmlns="http://www.w3.org/2000/svg"
  >
    <path
      d="M0 0h24v24H0z"
      fill="none"
      stroke="none"
    />
    <path
      d="M5 12l5 5l10 -10"
    />
  </svg>
</div>
`;
